flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaWriter.java

```diff
@@ -0,0 +1,379 @@
+        producers.add(currentProducer);
+        return committables;
+    }
+
+    @Override
+    public List<KafkaWriterState> snapshotState() throws IOException {
+        return ImmutableList.of(kafkaWriterState);
+    }
+
+    @Override
+    public void close() throws Exception {
+        currentProducer.close(Duration.ZERO);
+    }
+
+    private KafkaWriterState recoverAndInitializeState(List<KafkaWriterState> recoveredStates) {
+        final int subtaskId = kafkaSinkContext.getParallelInstanceId();
+        if (recoveredStates.isEmpty()) {
+            final KafkaWriterState state =
+                    new KafkaWriterState(transactionalIdPrefix, subtaskId, 0);
+            abortTransactions(getTransactionsToAbort(state, new ArrayList<>()));
+            return state;
+        }
+        final Map<Integer, KafkaWriterState> taskOffsetMapping =
+                recoveredStates.stream()
+                        .collect(
+                                Collectors.toMap(
+                                        KafkaWriterState::getSubtaskId, Function.identity()));
+        if (!taskOffsetMapping.containsKey(subtaskId)) {
+            throw new IllegalStateException(
+                    "It is expected that state from previous executions is distributed to the same subtask id.");
+        }
+        final KafkaWriterState lastState = taskOffsetMapping.get(subtaskId);
+        taskOffsetMapping.remove(subtaskId);
+        abortTransactions(
+                getTransactionsToAbort(lastState, new ArrayList<>(taskOffsetMapping.values())));
+        if (!lastState.getTransactionalIdPrefix().equals(transactionalIdPrefix)) {
+            LOG.warn(
+                    "Transactional id prefix from previous execution {} has changed to {}.",
+                    lastState.getTransactionalIdPrefix(),
+                    transactionalIdPrefix);
+            return new KafkaWriterState(transactionalIdPrefix, subtaskId, 0);
+        }
+        return new KafkaWriterState(
+                transactionalIdPrefix, subtaskId, lastState.getTransactionalIdOffset());
+    }
+
+    private void abortTransactions(Set<String> transactionsToAbort) {
+        transactionsToAbort.forEach(
+                transaction -> {
+                    // don't mess with the original configuration or any other
+                    // properties of the
+                    // original object
+                    // -> create an internal kafka producer on our own and do not rely
+                    // on
+                    //    initTransactionalProducer().
+                    final Properties myConfig = new Properties();
+                    myConfig.putAll(kafkaProducerConfig);
+                    myConfig.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, transaction);
+                    LOG.info("Aborting Kafka transaction {}.", transaction);
+                    FlinkKafkaInternalProducer<byte[], byte[]> kafkaProducer = null;
```

Comment:
We could use try-with-resource if we relay `close` to `close(0)` in `FlinkKafkaInternalProducer`.

Reply:
Hmm, currently the KafkaComitter leverages the try-with-resources logic because it is important to block until the transaction is finished. By changing this behavior we would need to change the committer try-with-resource block. I am not sure we gain much here.